### PR TITLE
feat: add SecretProviderClass metrics

### LINKED ISSUE #514: SecretProviderClass metrics
**Describe the solution you'd like**
I want to add metrics of generic status about `SecretProviderClass` respectively like kube-state-metrics' [Secret metrics](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/secret-metrics.md). It will be useful to summarize SecretProviderClass with labels.

**Anything else you would like to add:**
| Metric name| Metric type | Labels/tags |
| ---------- | ----------- | ----------- |
| kube_secretproviderclass_info | Gauge | `secretproviderclass`=&lt;spc-name&gt; <br> `namespace`=&lt;spc-namespace&gt; |
| kube_secretproviderclass_type | Gauge | `secretproviderclass`=&lt;spc-name&gt; <br> `namespace`=&lt;spc-namespace&gt; <br> `type`=&lt;syncing-secret-type&gt; <br> `provider`=&lt;provider-name&gt;|
| kube_secretproviderclass_labels | Gauge | `secretproviderclass`=&lt;spc-name&gt; <br> `namespace`=&lt;spc-namespace&gt; <br> `label_SPC_LABEL`=&lt;SPC_LABEL&gt; |
| kube_secretproviderclass_created  | Gauge | `secretproviderclass`=&lt;spc-name&gt; <br> `namespace`=&lt;spc-namespace&gt; |
| kube_secretproviderclass_metadata_resource_version  | Gauge | `secretproviderclass`=&lt;spc-name&gt; <br> `namespace`=&lt;spc-namespace&gt; |

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): `v0.0.21`
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by mitsutaka**
Signed-off-by: Mitz Amano <mitz@cloudflare.com>

**What this PR does / why we need it**:
This PR adds SecretProviderClass metrics that start with `kube_secretproviderclass_`.
Also upgrade `otel` packages to latest versions

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #514 

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/metrics/stats_reporter.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+var (
+	secretProviderClassKey    = "secretproviderclass"
+	namespaceKey              = "namespace"
+	secretObjectName          = "secret_name"
+	secretObjectTypeKeyPrefix = "secret_type_"
+	providerKey               = "provider"
+	labelKeyPrefix            = "label_"
+
```
Comment (aramase) [inline pkg/metrics/stats_reporter.go:41]: I think we should also add the os type to the metrics

https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/dcf491d1c4b3fee8cec6320d2d9224cfae57cc2c/pkg/secrets-store/stats_reporter.go#L28

File: pkg/metrics/stats_reporter.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	metricInfoName                    = "kube_secretproviderclass_info"
+	metricInfoDesc                    = "Information about SecretProviderClass"
+	metricTypeName                    = "kube_secretproviderclass_type"
+	metricTypeDesc                    = "Type about SecretProviderClass"
+	metricLabelsName                  = "kube_secretproviderclass_labels"
+	metricLabelsDesc                  = "Kubernetes labels converted to OpenTelemetry labels"
+	metricCreatedName                 = "kube_secretproviderclass_created"
+	metricCreatedDesc                 = "Unix creation timestamp"
+	metricMetadataResourceVersionName = "kube_secretproviderclass_metadata_resource_version"
```
Comment (aramase) [inline pkg/metrics/stats_reporter.go:50]: What is the significance of this metric?

File: pkg/metrics/stats_reporter.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	if metricCreated, err = obs.NewFloat64ValueObserver(metricCreatedName, metric.WithDescription(metricCreatedDesc)); err != nil {
+		return err
+	}
+
+	if metricMetadataResourceVersion, err = obs.NewFloat64ValueObserver(metricMetadataResourceVersionName, metric.WithDescription(metricMetadataResourceVersionDesc)); err != nil {
+		return err
+	}
+
```
Comment (aramase) [inline pkg/metrics/stats_reporter.go:161]: Could you also add unit tests for this package?

File: pkg/metrics/stats_reporter.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/metric"
+	"go.opentelemetry.io/otel/label"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
+)
+
+var (
```
Comment (tam7t) [inline pkg/metrics/stats_reporter.go:34]: maybe `const` instead of `var`?

File: pkg/metrics/stats_reporter.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	metricCreatedDesc                 = "Unix creation timestamp"
+	metricMetadataResourceVersionName = "kube_secretproviderclass_metadata_resource_version"
+	metricMetadataResourceVersionDesc = "Resource version representing a specific version of secret"
+)
+
+type reporter struct {
+	client.Client
+}
+
+func NewSecretProviderClassReporter(client client.Client) manager.Runnable {
```
Comment (tam7t) [inline pkg/metrics/stats_reporter.go:58]: Request a doc comment

File: pkg/metrics/stats_reporter.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	meter := global.Meter("secretsstore")
+	var (
+		err                           error
+		metricInfo                    metric.Int64ValueObserver
+		metricType                    metric.Int64ValueObserver
+		metricLabel                   metric.Int64ValueObserver
+		metricCreated                 metric.Float64ValueObserver
+		metricMetadataResourceVersion metric.Float64ValueObserver
+		// lock prevents a race between batch observer and instrument registration.
+		lock sync.Mutex
```
Comment (tam7t) [inline pkg/metrics/stats_reporter.go:90]: arn't these scoped to only the individual `report()` calls? Should they be members of the `reporter` struct instead?

File: pkg/metrics/stats_reporter.go (hunk 2 above)
Comment (mitsutaka) [reply pkg/metrics/stats_reporter.go]: this is based on kube-state-metrics. I don't think it is essential metrics for now. deleted.

File: pkg/metrics/stats_reporter.go (hunk 1 above)
Comment (mitsutaka) [reply pkg/metrics/stats_reporter.go:41]: added

File: pkg/metrics/stats_reporter.go (hunk 4 above)
Comment (mitsutaka) [reply pkg/metrics/stats_reporter.go]: fixed

File: pkg/metrics/stats_reporter.go (hunk 5 above)
Comment (mitsutaka) [reply pkg/metrics/stats_reporter.go:65]: Added

File: pkg/metrics/stats_reporter.go (hunk 6 above)
Comment (mitsutaka) [reply pkg/metrics/stats_reporter.go]: moved them to struct.

**Comment by mitsutaka** [comment]
@aramase It should be in review now

**Comment by yeya24** [comment]
@mitsutaka Hello, just a quick question. Why do u choose OTEL metrics rather than https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics?

**Comment by mitsutaka** [comment]
> @mitsutaka Hello, just a quick question. Why do u choose OTEL metrics rather than https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics?

Because it originally expects using OTEL https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/docs/book/src/topics/metrics.md `github.com/prometheus/client_golang/prometheus` is fine to me but I don't want to use multiple packages for metrics.

**Comment by yeya24** [comment]
> > @mitsutaka Hello, just a quick question. Why do u choose OTEL metrics rather than https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics?
> 
> 
> 
> Because it originally expects using OTEL https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/docs/book/src/topics/metrics.md `github.com/prometheus/client_golang/prometheus` is fine to me but I don't want to use multiple packages for metrics.

Oh, I am new to this project. Thanks for pointing it out!

File: pkg/metrics/stats_reporter.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	r.metricLabel = obs.NewInt64ValueObserver(metricLabelsName, metric.WithDescription(metricLabelsDesc))
+	r.metricCreated = obs.NewFloat64ValueObserver(metricCreatedName, metric.WithDescription(metricCreatedDesc))
+}
+
+func (r *secretProviderClassReporter) batchFunc() metric.BatchObserverFunc {
+	return func(ctx context.Context, result metric.BatchObserverResult) {
+		r.lock.Lock()
+		defer r.lock.Unlock()
+
+		spcList := &v1alpha1.SecretProviderClassList{}
```
Comment (aramase) [inline pkg/metrics/stats_reporter.go:89]: How frequently is this list call done?

File: go.mod [hunk 8]
```diff
@@ -5,14 +5,18 @@ go 1.16
...
-	github.com/google/go-cmp v0.5.5
+	github.com/google/go-cmp v0.5.6
 	github.com/kubernetes-csi/csi-lib-utils v0.7.1
 	github.com/kubernetes-csi/csi-test/v4 v4.0.2
 	github.com/onsi/gomega v1.13.0
 	github.com/prometheus/client_golang v1.11.0
 	github.com/stretchr/testify v1.7.0
-	go.opentelemetry.io/otel v0.13.0
-	go.opentelemetry.io/otel/exporters/metric/prometheus v0.13.0
+	go.opentelemetry.io/otel v1.0.0-RC1
```
[the remark this answers is not in the document]
Comment (tam7t) [reply go.mod:14]: I've run into this where some tooling will auto-try to pickup the RC package, but I'd feel more comfortable if we tagged against `v0.20.0` which is the 'latest' according to https://pkg.go.dev/go.opentelemetry.io/otel

**Comment by aramase** [comment]
@mitsutaka  Could you take a look at the comment and also rebase the PR? If this PR is not being actively worked on I would recommend we `/close` this and reopen when ready to review again.

**Comment by aramase** [comment]
Closing this PR now. Please feel free to reopen when it's ready for review again.

/close

**Pull request closed without merging**